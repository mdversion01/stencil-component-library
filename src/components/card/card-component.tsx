import { Component, Prop, Event, EventEmitter, h } from '@stencil/core';

@Component({
  tag: 'card-component',
  styleUrl: 'card.scss',
  shadow: false,
})
export class CardComponent {
  @Prop() actions = false;
  @Prop() altText = '';
  @Prop() ariaLabel: string = 'Card section';
  @Prop() cardMaxWidth = '20';
  @Prop() classNames = '';
  @Prop() elevation = '';
  @Prop() img = false;
  @Prop() imgSrc = '';
  @Prop() inlineStyles = '';
  @Prop() noFooter = false;
  @Prop() noHeader = false;
  @Prop() imgHeight = '11.25rem';
  @Prop() tab?: string;

  @Event() customClick: EventEmitter;

  private parseStyleString(styleStr: string): { [key: string]: string } {
    return styleStr.split(';').map(s => s.trim()).filter(s => s).reduce((acc, item) => {
      const [key, value] = item.split(':').map(x => x.trim());
      if (key && value) {
        acc[key] = value;
      }
      return acc;
    }, {} as { [key: string]: string });
  }

  // private handleKeyDown(event: KeyboardEvent) {
  //   if ((event.key === 'Enter' || event.key === ' ') && this.tabIndexs !== undefined) {
  //     event.preventDefault();
  //     this.customClick.emit();
  //   }
  // }

  private handleKeyDown(event: KeyboardEvent) {
    if (event.key === 'Enter' || event.key === ' ') {
      this.customClick.emit();
    }
  }

  render() {
    const baseClasses = ['card', this.classNames];
    if (this.elevation) baseClasses.push(`elevated-${this.elevation}`);

    const classAttribute = baseClasses.filter(Boolean).join(' ');

    const combinedStyle = `${this.inlineStyles}; max-width: ${this.cardMaxWidth}rem;`;
    const divStyle = this.parseStyleString(combinedStyle);

    const imgStyles = {
      height: this.imgHeight,
      width: '100%',
      display: 'block',
    };

    return (
      <div
        class={classAttribute}
        style={divStyle}
        role="region"
        aria-label={this.ariaLabel}
        {...(this.tab ? { tabindex: this.tab } : {})}
        onKeyDown={(e) => this.handleKeyDown(e)}
      >
        {this.img && (
          <img
            src={this.imgSrc}
            class="card-img-top"
            alt={this.altText || 'Card image'}
            style={imgStyles}
          />
        )}
        {!this.noHeader && (
          <div class="card-header">
            <slot name="header" />
          </div>
        )}
        <div class="card-body">
          <h5 class="card-title">
            <slot name="title" />
          </h5>
          <div class="card-text">
            <slot name="text" />
          </div>
        </div>
        {this.actions && (
          <div class="card-actions">
            <slot name="actions" />
          </div>
        )}
        {!this.noFooter && (
          <div class="card-footer">
            <slot name="footer" />
          </div>
        )}
      </div>
    );
  }
}
