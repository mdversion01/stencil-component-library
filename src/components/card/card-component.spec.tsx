import { newSpecPage } from '@stencil/core/testing';
import { CardComponent } from './card-component';

describe('card-component', () => {
  it('renders basic card with default aria attributes', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component></card-component>`,
    });
    const div = page.root.querySelector('div');
    expect(div.getAttribute('role')).toBe('region');
    expect(div.getAttribute('aria-label')).toBe('Card section');
    expect(page.root).toMatchSnapshot();
  });

  it('renders card with custom aria-label', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component aria-label="Profile Card"></card-component>`,
    });
    const div = page.root.querySelector('div');
    expect(div.getAttribute('aria-label')).toBe('Profile Card');
    expect(page.root).toMatchSnapshot();
  });

  it('renders card with header and footer slots', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `
        <card-component>
          <div slot="header">Header Content</div>
          <div slot="title">Card Title</div>
          <div slot="text">Main text content</div>
          <div slot="footer">Footer Content</div>
        </card-component>
      `,
    });
    expect(page.root.querySelector('.card-header')).toBeTruthy();
    expect(page.root.querySelector('.card-footer')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('renders card with image', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component img img-src="https://example.com/image.jpg" alt-text="Sample Image"></card-component>`,
    });
    const img = page.root.querySelector('img');
    expect(img).toBeTruthy();
    expect(img.getAttribute('alt')).toBe('Sample Image');
    expect(page.root).toMatchSnapshot();
  });

  it('renders card with actions slot', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `
        <card-component actions>
          <div slot="actions">Action Button</div>
        </card-component>
      `,
    });
    expect(page.root.querySelector('.card-actions')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('applies elevation and custom classes', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component elevation="2" class-names="custom-card"></card-component>`,
    });
    const div = page.root.querySelector('div');
    expect(div.classList.contains('elevated-2')).toBe(true);
    expect(div.classList.contains('custom-card')).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('is focusable when tabindex is provided', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component tab="0"></card-component>`,
    });
    const div = page.root.querySelector('div');
    expect(div.getAttribute('tabindex')).toBe('0');
    expect(page.root).toMatchSnapshot();
  });

  it('responds to Enter and Space key events if clickable', async () => {
    const page = await newSpecPage({
      components: [CardComponent],
      html: `<card-component tab="0"></card-component>`,
    });

    const clickSpy = jest.fn();
    page.root.addEventListener('customClick', clickSpy);

    const div = page.root.querySelector('div');
    div.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));
    div.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));
    await page.waitForChanges();

    expect(clickSpy).toHaveBeenCalledTimes(2);
  });

});
