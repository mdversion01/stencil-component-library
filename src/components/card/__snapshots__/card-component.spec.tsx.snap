// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card-component applies elevation and custom classes 1`] = `
<card-component class-names="custom-card" elevation="2">
  <!---->
  <div aria-label="Card section" class="card custom-card elevated-2" role="region" style="max-width: 20rem;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;

exports[`card-component is focusable when tabindex is provided 1`] = `
<card-component tab="0">
  <!---->
  <div aria-label="Card section" class="card" role="region" tabindex="0" style="max-width: 20rem;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;

exports[`card-component renders basic card with default aria attributes 1`] = `
<card-component>
  <!---->
  <div aria-label="Card section" class="card" role="region" style="max-width: 20rem;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;

exports[`card-component renders card with actions slot 1`] = `
<card-component actions="">
  <!---->
  <div aria-label="Card section" class="card" role="region" style="max-width: 20rem;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-actions">
      <div slot="actions">
        Action Button
      </div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;

exports[`card-component renders card with custom aria-label 1`] = `
<card-component aria-label="Profile Card">
  <!---->
  <div aria-label="Profile Card" class="card" role="region" style="max-width: 20rem;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;

exports[`card-component renders card with header and footer slots 1`] = `
<card-component>
  <!---->
  <div aria-label="Card section" class="card" role="region" style="max-width: 20rem;">
    <div class="card-header">
      <div slot="header">
        Header Content
      </div>
    </div>
    <div class="card-body">
      <h5 class="card-title">
        <div slot="title">
          Card Title
        </div>
      </h5>
      <div class="card-text">
        <div slot="text">
          Main text content
        </div>
      </div>
    </div>
    <div class="card-footer">
      <div slot="footer">
        Footer Content
      </div>
    </div>
  </div>
</card-component>
`;

exports[`card-component renders card with image 1`] = `
<card-component alt-text="Sample Image" img="" img-src="https://example.com/image.jpg">
  <!---->
  <div aria-label="Card section" class="card" role="region" style="max-width: 20rem;">
    <img alt="Sample Image" class="card-img-top" src="https://example.com/image.jpg" style="height: 11.25rem; width: 100%; display: block;">
    <div class="card-header"></div>
    <div class="card-body">
      <h5 class="card-title"></h5>
      <div class="card-text"></div>
    </div>
    <div class="card-footer"></div>
  </div>
</card-component>
`;
